File: python3/first_bad_version.py
"""
Space   : O(1)
Time    : O(log n)
"""

# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):


class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        low = 1
        high = n

        while(low <= high):
            mid = (low + high) // 2
            if not isBadVersion(mid):
                if low != mid:
                    low = mid
                else:
                    low = mid + 1
            elif isBadVersion(mid):
                if not isBadVersion(mid-1) and isBadVersion(mid):
                    return mid
                else:
                    high = mid

File: golang/first_bad_version.go
package main

// Space	: O(1)
// Time		: O(log n)

/**
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */

func firstBadVersion(n int) int {
	left, right := 0, n
	mid := (left + right) / 2

	for left <= right {
		mid = (left + right) / 2
		if !isBadVersion(mid) {
			if left != mid {
				left = mid
			} else {
				left = mid + 1
			}
		} else {
			if !isBadVersion(mid-1) && isBadVersion(mid) {
				break
			} else {
				right = mid
			}
		}
	}
	return mid
}

// mock function
func isBadVersion(n int) bool {
	return n == n
}
